react/the_react_ecosystem/react_router.md

```diff
@@ -0,0 +1,589 @@
+### Introduction
+
+Up until this point in the curriculum, we have been building one-page applications. However, for any larger scale application, we are going to have multiple pages. Thankfully, the browser allows client-side Javascript to manage the way a user can navigate, with the [History API](https://developer.mozilla.org/en-US/docs/Web/API/History_API/Working_with_the_History_API). We can leverage the power of this to manage routing in React with the help of a package like `react-router-dom`.
+
+### Lesson Overview
+
+This section contains a general overview of topics that you will learn in this lesson.
+
+- What is client-side routing?
+- How do you use React Router to do client-side routing?
+- How do you create nested and dynamic paths?
+- How do you add a "catch-all" route?
+- How do you add protected routes?
+
+### Client-Side Routing
+
+Client-side routing is the type of routing where Javascript takes over the duty of handling the routes in an application. Client-side routing helps in building single-page applications (SPAs) without refreshing as the user navigates. For example, when a user clicks a navbar element, the URL changes and the view of the page is modified accordingly, within the client.
+
+Say you are cooking some chicken. If you want to cook it well and nice, you will have to:
+
+1. Put the chicken in the microwave and set it to cook with appropriate time and heating
+2. Wait till the microwave gives out that satisfying **ding**
+3. Start munching!
+
+This is common to all websites, you set the microwave up for what you want (visit any URL, like https://theodinproject.com/), wait for the microwave to be done with the cooking (the loading screen), and tada, enjoy your delicious food (your page is ready for use). But what if you forgot to add some spices before you cooked it up? You have to repeat this flow again:
+
+1. Add the spices to the chicken
+2. Put it back into the microwave and set it up to be reheated
+3. Wait for it to be nice and warm
+4. Now you can eat it!
+
+Here is where we reiterate, **the chicken needs to be reheated**. In a general multi-page application (MPAs), the browser reloads every time you click on a link to navigate. With client-side routing, **you never leave the page you are on** - you bring the microwave to the table to ensure that you don't run into the "missing spices" issues. The data is rendered with Javascript, instead of letting the browser handle it.
+
+### A Reactive Solution
+
+While this allows for nicer, app-like interactions (since you are controlling the routes, you can make fancy css animations across route changes), a lot of caveats can be missed. Browsers reloads notify screen-readers of new content to read, so you will need to notify screen-readers of route updates manually. However, with the help of a robust library, you can often address these concerns!
```

Comment:
Wording suggestion:
```suggestion
While client-side routing allows for nicer, app-like interactions (since you are controlling the routes, you can make fancy css animations across route changes), a lot of caveats can be missed. Browsers reloads notify screen-readers of new content to read, so you will need to notify screen-readers of route updates manually. However, with the help of a robust library, you can often address these concerns!
```